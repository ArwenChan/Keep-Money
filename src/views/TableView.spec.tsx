import {
  fireEvent,
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
  within,
} from '@testing-library/react'
import React, { createRef, Suspense } from 'react'
require('fake-indexeddb/auto')

describe('test table view', () => {
  it('test table view render', async () => {
    const { AppContext } = await import('../App')
    const { default: TableView } = await import('./TableView')
    const appRef = createRef<HTMLDivElement>()
    let container: HTMLElement
    await waitFor(() => {
      container = render(
        <div className="App" id="app" ref={appRef}>
          <AppContext.Provider value={appRef}>
            <Suspense fallback={'loading...'}>
              <TableView />
            </Suspense>
          </AppContext.Provider>
        </div>
      ).container
    })
    let items: HTMLElement[]
    expect(screen.queryByText('loading...')).toBeInTheDocument()
    await waitFor(() => {
      items = screen.getAllByRole('item')
      expect(items.length).toEqual(8)
    })
    //@ts-ignore
    expect(container).toMatchSnapshot()
    expect(screen.getByText('更多')).toBeInTheDocument()
    const button = screen.getByTestId('add-button')
    fireEvent.click(button)
    // into dialog
    expect(screen.getByText('添加记账')).toBeInTheDocument()
    const dialog = within(screen.getByRole('dialog'))
    const input = dialog.getByLabelText('金额') as HTMLInputElement

    const buttons = dialog.queryAllByRole('button')
    expect(buttons.length).toEqual(3)
    fireEvent.mouseDown(buttons[0])
    let lis = screen.queryAllByRole('option')
    expect(lis.length).toEqual(13)
    fireEvent.click(lis[1])
    await waitForElementToBeRemoved(lis)
    fireEvent.change(input, {
      target: {
        value: '123',
      },
    })
    fireEvent.click(buttons[2])
    // dialog remove
    await waitForElementToBeRemoved(buttons)
    // data added
    await waitFor(() => {
      expect(screen.queryAllByRole('item').length).toEqual(9)
    })

    const calendar = screen.getByTestId('calendar')
    expect(calendar).toBeInTheDocument()
    fireEvent.change(within(calendar).getByRole('month'), {
      target: {
        value: '2010-07',
      },
    })
    await waitForElementToBeRemoved(items!)

    fireEvent.change(within(calendar).getByRole('month'), {
      target: {
        value: '2019-11',
      },
    })
    await waitFor(() => {
      items = screen.getAllByRole('item')
      expect(items.length).toEqual(2)
    })
    //@ts-ignore
    fireEvent.click(within(items[0]).getByRole('checkbox'))
    expect(screen.queryByTestId('calendar')).toBeNull()
    expect(screen.queryByText('1 selected')).not.toBeNull()
    fireEvent.click(screen.getByLabelText('delete'))
    // @ts-ignore
    await waitForElementToBeRemoved(items[0])
    items = screen.getAllByRole('item')
    expect(items.length).toEqual(1)

    fireEvent.mouseDown(screen.getByText('分类'))
    const options = screen.queryAllByRole('option')
    expect(options.length).toEqual(13)
    expect(options[0].textContent).toBe('收入')
    expect(options[1].textContent).toBe('股票投资')
    fireEvent.click(screen.getByText('股票投资'))
    await waitFor(() => {
      expect(screen.queryAllByRole('item').length).toEqual(0)
    })
  }, 20000)
})
