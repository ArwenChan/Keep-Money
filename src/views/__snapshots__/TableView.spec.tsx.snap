// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test table view test table view render 1`] = `
<div>
  <div
    class="App"
    id="app"
  >
    <div
      class="view"
    >
      <div
        class="root"
      >
        <div
          class="MuiPaper-root paper MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular toolbarRoot MuiToolbar-gutters"
          >
            <div
              class="MuiTypography-root toolbarTitle MuiTypography-h6"
              id="tableTitle"
            >
              Keep Money
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root"
              data-testid="calendar"
            >
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input"
                  id="date"
                  role="month"
                  type="month"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="MuiTableContainer-root"
          >
            <table
              aria-label="bill"
              aria-labelledby="tableTitle"
              class="MuiTable-root table"
            >
              <thead
                class="MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-paddingCheckbox"
                    scope="col"
                  >
                    <span
                      aria-disabled="false"
                      class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <input
                          aria-label="select all"
                          class="PrivateSwitchBase-input-4"
                          data-indeterminate="false"
                          type="checkbox"
                          value=""
                        />
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </span>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft"
                    scope="col"
                  >
                    <div
                      class="MuiFormControl-root"
                    >
                      <div
                        class="MuiInputBase-root MuiInput-root MuiInput-underline filter MuiInputBase-formControl MuiInput-formControl"
                        data-testid="mutiple-checkbox-filter"
                      >
                        <div
                          aria-haspopup="listbox"
                          aria-labelledby="mutiple-checkbox-filter-label mutiple-checkbox-filter"
                          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                          id="mutiple-checkbox-filter"
                          role="button"
                          tabindex="0"
                        >
                          <p
                            class="render"
                          >
                            分类
                          </p>
                        </div>
                        <input
                          aria-hidden="true"
                          class="MuiSelect-nativeInput"
                          tabindex="-1"
                          value=""
                        />
                        <svg
                          class="MuiSelect-icon"
                        >
                          filter-outlined.svg
                        </svg>
                      </div>
                    </div>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                    scope="col"
                  >
                    <span
                      aria-disabled="false"
                      class="MuiButtonBase-root MuiTableSortLabel-root"
                      role="button"
                      tabindex="0"
                    >
                      金额
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                        />
                      </svg>
                    </span>
                  </th>
                  <th
                    aria-sort="ascending"
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                    scope="col"
                  >
                    <span
                      aria-disabled="false"
                      class="MuiButtonBase-root MuiTableSortLabel-root MuiTableSortLabel-active"
                      role="button"
                      tabindex="0"
                    >
                      时间
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                        />
                      </svg>
                    </span>
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root"
              >
                <tr
                  aria-checked="false"
                  class="MuiTableRow-root MuiTableRow-hover"
                  role="item"
                  tabindex="-1"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
                  >
                    <span
                      aria-disabled="false"
                      class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <input
                          aria-labelledby="enhanced-table-checkbox-0"
                          class="PrivateSwitchBase-input-4"
                          data-indeterminate="false"
                          type="checkbox"
                          value=""
                        />
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </span>
                  </td>
                  <th
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    id="enhanced-table-checkbox-0"
                    role="cell"
                    scope="row"
                  >
                    房贷
                  </th>
                  <td
                    class="MuiTableCell-root MuiTableCell-body outcomeMoney MuiTableCell-alignRight"
                  >
                    5400
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    2019-11-29
                  </td>
                </tr>
                <tr
                  aria-checked="false"
                  class="MuiTableRow-root MuiTableRow-hover"
                  role="item"
                  tabindex="-1"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
                  >
                    <span
                      aria-disabled="false"
                      class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <input
                          aria-labelledby="enhanced-table-checkbox-1"
                          class="PrivateSwitchBase-input-4"
                          data-indeterminate="false"
                          type="checkbox"
                          value=""
                        />
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </span>
                  </td>
                  <th
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    id="enhanced-table-checkbox-1"
                    role="cell"
                    scope="row"
                  >
                    日常饮食
                  </th>
                  <td
                    class="MuiTableCell-root MuiTableCell-body outcomeMoney MuiTableCell-alignRight"
                  >
                    5000
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    2019-11-30
                  </td>
                </tr>
                <tr
                  aria-checked="false"
                  class="MuiTableRow-root MuiTableRow-hover"
                  role="item"
                  tabindex="-1"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
                  >
                    <span
                      aria-disabled="false"
                      class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <input
                          aria-labelledby="enhanced-table-checkbox-2"
                          class="PrivateSwitchBase-input-4"
                          data-indeterminate="false"
                          type="checkbox"
                          value=""
                        />
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </span>
                  </td>
                  <th
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    id="enhanced-table-checkbox-2"
                    role="cell"
                    scope="row"
                  >
                    房屋租赁
                  </th>
                  <td
                    class="MuiTableCell-root MuiTableCell-body outcomeMoney MuiTableCell-alignRight"
                  >
                    1500
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    2019-12-1
                  </td>
                </tr>
                <tr
                  aria-checked="false"
                  class="MuiTableRow-root MuiTableRow-hover"
                  role="item"
                  tabindex="-1"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
                  >
                    <span
                      aria-disabled="false"
                      class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <input
                          aria-labelledby="enhanced-table-checkbox-3"
                          class="PrivateSwitchBase-input-4"
                          data-indeterminate="false"
                          type="checkbox"
                          value=""
                        />
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </span>
                  </td>
                  <th
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    id="enhanced-table-checkbox-3"
                    role="cell"
                    scope="row"
                  >
                    日常饮食
                  </th>
                  <td
                    class="MuiTableCell-root MuiTableCell-body outcomeMoney MuiTableCell-alignRight"
                  >
                    4000
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    2019-12-26
                  </td>
                </tr>
                <tr
                  aria-checked="false"
                  class="MuiTableRow-root MuiTableRow-hover"
                  role="item"
                  tabindex="-1"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
                  >
                    <span
                      aria-disabled="false"
                      class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <input
                          aria-labelledby="enhanced-table-checkbox-4"
                          class="PrivateSwitchBase-input-4"
                          data-indeterminate="false"
                          type="checkbox"
                          value=""
                        />
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </span>
                  </td>
                  <th
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    id="enhanced-table-checkbox-4"
                    role="cell"
                    scope="row"
                  >
                    基金投资
                  </th>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    1000
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    2019-12-26
                  </td>
                </tr>
                <tr
                  aria-checked="false"
                  class="MuiTableRow-root MuiTableRow-hover"
                  role="item"
                  tabindex="-1"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
                  >
                    <span
                      aria-disabled="false"
                      class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <input
                          aria-labelledby="enhanced-table-checkbox-5"
                          class="PrivateSwitchBase-input-4"
                          data-indeterminate="false"
                          type="checkbox"
                          value=""
                        />
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </span>
                  </td>
                  <th
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    id="enhanced-table-checkbox-5"
                    role="cell"
                    scope="row"
                  >
                    工资
                  </th>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    3000
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    2019-12-26
                  </td>
                </tr>
                <tr
                  aria-checked="false"
                  class="MuiTableRow-root MuiTableRow-hover"
                  role="item"
                  tabindex="-1"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
                  >
                    <span
                      aria-disabled="false"
                      class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <input
                          aria-labelledby="enhanced-table-checkbox-6"
                          class="PrivateSwitchBase-input-4"
                          data-indeterminate="false"
                          type="checkbox"
                          value=""
                        />
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </span>
                  </td>
                  <th
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    id="enhanced-table-checkbox-6"
                    role="cell"
                    scope="row"
                  >
                    基金投资
                  </th>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    2000
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    2019-12-26
                  </td>
                </tr>
                <tr
                  aria-checked="false"
                  class="MuiTableRow-root MuiTableRow-hover"
                  role="item"
                  tabindex="-1"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox"
                  >
                    <span
                      aria-disabled="false"
                      class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <input
                          aria-labelledby="enhanced-table-checkbox-7"
                          class="PrivateSwitchBase-input-4"
                          data-indeterminate="false"
                          type="checkbox"
                          value=""
                        />
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </span>
                  </td>
                  <th
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    id="enhanced-table-checkbox-7"
                    role="cell"
                    scope="row"
                  >
                    股票投资
                  </th>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    2000
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                  >
                    2019-12-26
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
      <div
        class="bar"
      >
        <span
          class="barSummary"
          data-testid="bar-string"
        >
          收入：
          8000
          ，
          <span
            class="red"
          >
            支出：
            15900
          </span>
          <span
            class="underline"
          >
            更多
          </span>
        </span>
        <button
          class="MuiButtonBase-root MuiIconButton-root button"
          data-testid="add-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root icon"
              focusable="false"
              style="font-size: 60px; color: black;"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
